### DISCUSSION TITLE
Why are there disconnects seemingly at random?

### DISCUSSION THREAD

**Discussion by Zitronenjoghurt** [Rapptz/discord.py#9722] [General] [answered]
I have recently looked into my bots logs and noticed that theres 2 scenarios where it reconnects.

The first one seems to be intended as it sends a reconnect opcode.
```
discord.gateway For Shard ID None: WebSocket Event: {'t': None, 's': None, 'op': 7, 'd': None}
discord.gateway Received RECONNECT opcode.
```

But sometimes theres also this:
```
discord.gateway Received WSMessage(type=<WSMsgType.CLOSED: 257>, data=None, extra=None)
```
after which theres usually an error like:
```
discord.gateway Websocket closed with WSCloseCode.OK, cannot reconnect.
discord.client Dispatching event disconnect
discord.client Attempting a reconnect in 1.64s
Traceback (most recent call last):
  File "/home/user/Bots/BotName/venv/lib/python3.10/site-packages/discord/client.py", line 659, in connect
    await self.ws.poll_event()
  File "/home/user/Bots/BotName/venv/lib/python3.10/site-packages/discord/gateway.py", line 646, in poll_event
    raise ConnectionClosed(self.socket, shard_id=self.shard_id, code=code) from None
discord.errors.ConnectionClosed: Shard ID None WebSocket closed with 1000
```

It does automatically reconnect afterwards but Im just wondering what the reason is for discord to close the connection without asking for a reconnect. Is it a normal thing to happen? In the end its handled well and you dont notice it, though it still throws an error which shouldnt normally happen on a regular basis.

**Comment by Rapptz** [comment, marked as answer]
This is just part of the connection flow. Discord disconnects clients every 15 minutes to 2 hours for load balancing reasons. Some of those rarely get invalidated and you have to do a full reconnect but most of the time it's just resumed properly.

**Comment by 1337faye** [comment]
It seems that it is possible to bypass this by sending dummy audio data at least in my testing. Without it, the bot never lasts more than 4 hours, whilst now after some tweaks it has been up for 7 hours so far.

I used this function.

```python
 async def voice_keepalive(self, voice_client):
        """Keeps the voice connection alive by periodically sending silence packets"""
        print("Starting voice keepalive task")
        while True:
            if voice_client.is_connected():
                # Send a silence packet every 15 seconds
                voice_client.send_audio_packet(b'\xF8\xFF\xFE', encode=False)
                await asyncio.sleep(15)
            else:
                break
```

Then I used Asyncio (python library) to make this a task that runs asynchronously. Here's an example of executing the task after the call begins. Note that I'm using discord self (a fork of discord.py) since my selfbot is supposed to be used in my group chats, the implementation for a bot would be very similar.

```python
self.keepalive_task = None # ideally, you would have this initialized once when the script first starts.

voice_client = await group_channel.connect()
await message.channel.send(f'** User "{self.user.name}" has successfully joined the call.**')
# Start the keepalive task
if self.keepalive_task:
    self.keepalive_task.cancel()
self.keepalive_task = asyncio.create_task(self.voice_keepalive(voice_client))
```

Very late but I hope this helps.
